include/common/core/freertos_task.hpp

```diff
@@ -8,31 +8,86 @@
 
 namespace freertos_task {
 
-template <uint32_t StackDepth, UBaseType_t Priority>
+template <uint32_t StackDepth>
+struct FreeRTOSTaskControl {
+    FreeRTOSTaskControl() = default;
+    auto operator=(FreeRTOSTaskControl&) -> FreeRTOSTaskControl& = delete;
+    auto operator=(FreeRTOSTaskControl&&) -> FreeRTOSTaskControl&& = delete;
+    FreeRTOSTaskControl(FreeRTOSTaskControl&) = delete;
+    FreeRTOSTaskControl(FreeRTOSTaskControl&&) = delete;
+    ~FreeRTOSTaskControl() = default;
+
+    TaskHandle_t handle{};
+    StaticTask_t static_task{};
+    std::array<StackType_t, StackDepth> backing{};
+};
+
+/**
+ * A FreeRTOS task.
+ *
+ * @tparam StackDepth The size of the stack.
+ * @tparam EntryPoint The entry point type. Must be a copyable callable that
+ * takes no arguments.
+ */
+template <uint32_t StackDepth, typename EntryPoint>
 class FreeRTOSTask {
   public:
-    using EntryPoint = std::function<void()>;
-    FreeRTOSTask(const char* task_name, EntryPoint entry)
-        : entry{std::move(entry)} {
-        handle = xTaskCreateStatic(f, task_name, StackDepth, this, Priority,
-                                   backing.data(), &static_task);
+    /**
+     * Constructor
+     * @param entry  This is the entry point to the task.
+     * @param control The task control block.
+     * @param priority The tasks priority to use
+     * @param task_name Task name.
+     */
+    FreeRTOSTask(EntryPoint&& entry, FreeRTOSTaskControl<StackDepth>& control,
+                 UBaseType_t priority, const char* task_name)
+        : entry{std::move(entry)}, control{control} {
+        start(priority, task_name);
     }
-    auto operator=(FreeRTOSTask&) -> FreeRTOSTask& = delete;
-    auto operator=(FreeRTOSTask&&) -> FreeRTOSTask&& = delete;
-    FreeRTOSTask(FreeRTOSTask&) = delete;
-    FreeRTOSTask(FreeRTOSTask&&) = delete;
 
-    ~FreeRTOSTask() { vTaskDelete(handle); }
+    /**
+     * Constructor
+     * @param entry  This is the entry point to the task.
+     * @param control The task control block.
+     * @param priority The tasks priority to use
+     * @param task_name Task name.
+     */
+    FreeRTOSTask(EntryPoint& entry, FreeRTOSTaskControl<StackDepth>& control,
+                 UBaseType_t priority, const char* task_name)
+        : entry{entry}, control{control} {
+        start(priority, task_name);
+    }
+    FreeRTOSTask(FreeRTOSTask&) noexcept = default;
```

Comment:
do we really want this to be copyable? i'm not sure if it will autogenerate this or not, but if we want the semantics of this class to be an RAII-style task where the constructor also starts the task, it really doesn't seem like it should be copyable, or if it is the copy indicates creating a new underlying task.